File: julia/ra.jl
module RA

export raquery, raread, rawrite

const version = v"0.0.1"

FLAG_BIG_ENDIAN = UInt64(1<<0)
MAX_BYTES = UInt64(1<<31)

TYPE_NUM_TO_NAME = Dict(
  0 => "user",
  1 => "Int",
  2 => "UInt",
  3 => "Float",
  4 => "Complex"
)
TYPE_NAME_TO_NUM = Dict(
  "user" => 0,
  Int8 => 1,
  Int16 => 1,
  Int32 => 1,
  Int64 => 1,
  Int128 => 1,
  UInt8 => 2,
  UInt16 => 2,
  UInt32 => 2,
  UInt64 => 2,
  UInt128 => 2,
  Float16 => 3,
  Float32 => 3,
  Float64 => 3,
  Complex32 => 4,
  Complex64 => 4,
  Complex128 => 4,
  Complex{Float32} => 4,
  Complex{Float64} => 4
  )

#  header is 40 + 8*ndims bytes long
type RAHeader
  flags::UInt64  # file properties, such as endianness and future capabilities
  eltype::UInt64   # enum representing the element type in the array
  elbyte::UInt64    # enum representing the element type in the array
  size::UInt64   # size of data in bytes (may be compressed: check 'flags')
  ndims::UInt64  # number of dimensions in array
  dims::Vector{UInt64}
end

function getheader(io::IOStream)
  flags, eltype, elbits, size, ndims = read(io, UInt64, 5)
  dims = read(io, UInt64, ndims)
  return RAHeader(flags,eltype,elbits,size,ndims,dims)
end

# print the file header as YAML
function raquery(path::AbstractString)
  q = AbstractString[]
  push!(q, "---\nname: $path")
  fd = open(path,"r")
  h = getheader(fd)
  close(fd)
  juliatype = eval(parse(TYPE_NUM_TO_NAME[h.eltype]*"$(h.elbyte*8)"))
  endian = (h.flags & FLAG_BIG_ENDIAN) != 0 ? "big" : "little"
  assert(endian == "little") # big not implemented yet
  push!(q, "endian: $endian")
  push!(q, "type: $juliatype")
  push!(q, "size: $(h.size)")
  push!(q, "dimension: $(h.ndims)")
  push!(q, "shape:")
  for j = 1:h.ndims
    push!(q, "  - $(h.dims[j])")
  end
  push!(q, "...")
  join(q, "\n")
end

function raread(path::AbstractString)
  fd = open(path, "r")
  h = getheader(fd)
  juliatype = eval(parse("$(TYPE_NUM_TO_NAME[h.eltype])$(h.elbyte*8)"))
  data = read(fd, juliatype, round(Int,h.size/sizeof(juliatype)))
  data = reshape(data, [Int64(_) for _ in h.dims]...)
  close(fd)
  return data
end

function rawrite{T,N}(a::Array{T,N}, path::AbstractString)
  flags = UInt64(0)
  if ENDIAN_BOM == 0x01020304
    flags |=  FLAG_BIG_ENDIAN
  end
  fd = open(path, "w")
  write(fd, flags,
    UInt64(TYPE_NAME_TO_NUM[T]),
    UInt64(sizeof(T)),
    UInt64(length(a)*sizeof(eltype(a))),
    UInt64(ndims(a)),
    UInt64[_ for _ in size(a)],
    a)
  close(fd)
end

end

File: python/ra.py
import struct
import numpy as np

FLAG_BIG_ENDIAN = 0x01
MAGIC_NUMBER = 8746397786917265778
dtype_kind_to_enum = {'i':1,'u':2,'f':3,'c':4}
dtype_enum_to_name = {0:'user',1:'int',2:'uint',3:'float',4:'complex'}

def read(filename):
    f = open(filename,'rb')
    h = getheader(f)
    if h['eltype'] == 0:
        print 'Unable to convert user data. Returning raw byte string.'
        return f.read(h['size'])
    else:
        d = '%s%d' % (dtype_enum_to_name[h['eltype']], h['elbyte']*8)
        data = np.fromstring(f.read(h['size']), dtype=np.dtype(d))
        return data.reshape(h['dims'])

def getheader(f):
    filemagic = f.read(8)
    h = dict()
    h['flags'] = struct.unpack('<Q',f.read(8))[0]
    h['eltype'] = struct.unpack('<Q',f.read(8))[0]
    h['elbyte'] = struct.unpack('<Q',f.read(8))[0]
    h['size'] = struct.unpack('<Q',f.read(8))[0]
    h['ndims'] = struct.unpack('<Q',f.read(8))[0]
    h['dims'] = []
    for d in range(h['ndims']):
        h['dims'].append(struct.unpack('<Q',f.read(8))[0])
    return h


def query(filename):
  q = "---\nname: %s\n" % filename
  fd = open(filename,'r')
  h = getheader(fd)
  fd.close()
  if h['flags'] & FLAG_BIG_ENDIAN:
    endian = 'big'
  else:
    endian = 'little'
  assert endian == 'little'  # big not implemented yet
  q += 'endian: %s\n' % endian
  q += 'type: %s%d\n' % (dtype_enum_to_name[h['eltype']], h['elbyte']*8)
  q += 'size: %d\n' % h['size']
  q += 'dimension: %d\n' % h['ndims']
  q += 'shape:\n'
  for d in h['dims']:
     q += '  - %d\n' % d
  q += '...'
  return q


def write(data, filename):
    f = open(filename,'wb')
    flags = 0
    if data.dtype.str[0] == '>':
        flags |= FLAG_BIG_ENDIAN
    try:
        eltype = dtype_kind_to_enum[data.dtype.kind]
    except KeyError:
        eltype = 0
    elbyte = data.dtype.itemsize
    size = data.size*elbyte
    ndims = len(data.shape)
    dims = data.shape
    dims = np.array([_ for _ in data.shape]).astype('uint64')
    f.write(struct.pack('<Q', MAGIC_NUMBER))
    f.write(struct.pack('<Q', flags))
    f.write(struct.pack('<Q', eltype))
    f.write(struct.pack('<Q', elbyte))
    f.write(struct.pack('<Q', size))
    f.write(struct.pack('<Q', ndims))
    f.write(dims.tobytes())
    f.write(data.tobytes())
    f.close()
